{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d3c9c93b-4a94-47e6-8f51-de34dc9cc042",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter a DNA or RNA sequence:   ACTCCGGACTTTGGCCTTAGCAGTAGTTAGTGTGGGAGGCTGGGAAGACTGGGAGCAGTCTCTTAAACAA AAGCAAAAGAATAAGCTTCGGGCGCTGTAGTACCTGCCAGCTTTCGCCACAGGAGCCTAGGCTGCCACCT GTAGGTCACTTGGGCTCCAGCTATGTGGCTGCCTCTTCTGCTGGGTGCCTTACTCTGGGCAGTGCTGTGG TTGCTCAGGGACCGGCAGAGCCTGCCCGCCAGCAATGCCTTTGTCTTCATCACCGGCTGTGACTCAGGCT TTGGGCGCCTTCTGGCACTGCAGCTGGACCAGAGAGGCTTCCGAGTCCTGGCCAGCTGCCTGACCCCCTC CGGGGCCGAGGACCTGCAGCGGGTGGCCTCCTCCCGCCTCCACACCACCCTGTTGGATATCACTGATCCC CAGAGCGTCCAGCAGGCAGCCAAGTGGGTGGAGATGCACGTTAAGGAAGCAGGGCTTTTTGGTCTGGTGA ATAATGCTGGTGTGGCTGGTATCATCGGACCCACACCATGGCTGACCCGGGACGATTTCCAGCGGGTGCT GAATGTGAACACAATGGGTCCCATCGGGGTCACCCTTGCCCTGCTGCCTCTGCTGCAGCAAGCCCGGGGC CGGGTGATCAACATCACCAGCGTCCTGGGTCGCCTGGCAGCCAATGGTGGGGGCTACTGTGTCTCCAAAT TTGGCCTGGAGGCCTTCTCTGACAGCCTGAGGCGGGATGTAGCTCATTTTGGGATACGAGTCTCCATCGT GGAGCCTGGCTTCTTCCGAACCCCTGTGACCAACCTGGAGAGTCTGGAGAAAACCCTGCAGGCCTGCTGG GCACGGCTGCCTCCTGCCACACAGGCCCACTATGGGGGGGCCTTCCTCACCAAGTACCTGAAAATGCAAC AGCGCATCATGAACCTGATCTGTGACCCGGACCTAACCAAGGTGAGCCGATGCCTGGAGCATGCCCTGAC TGCTCGACACCCCCGAACCCGCTACAGCCCAGGTTGGGATGCCAAGCTGCTCTGGCTGCCTGCCTCCTAC CTGCCAGCCAGCCTGGTGGATGCTGTGCTCACCTGGGTCCTTCCCAAGCCTGCCCAAGCAGTCTACTGAA TCCAGCCTTCCAGCAAGAGATTGTTTTTCAAGGACAAGGACTTTGATTTATTTCTGCCCCCACCCTGGTA CTGCCTGGTGCCTGCCACAAAATAAGCACTAACAAAAGTGTATTGTTTAAAAAATAAAAAGAAGGTGGGC AGAAATGTGCCCAGTGGAA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Translated Protein Sequence: TPDFGLSSS\n"
     ]
    }
   ],
   "source": [
    "def translate_sequence(sequence):\n",
    "    \n",
    "    codon_table= {\n",
    "        'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L',\n",
    "        'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',\n",
    "        'AUU': 'I', 'AUC': 'I', 'AUA': 'I', 'AUG': 'M',\n",
    "        'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',\n",
    "        'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',\n",
    "        'AGU': 'S', 'AGC': 'S', 'CCU': 'P', 'CCC': 'P',\n",
    "        'CCA': 'P', 'CCG': 'P', 'ACU': 'T', 'ACC': 'T',\n",
    "        'ACA': 'T', 'ACG': 'T', 'GCU': 'A', 'GCC': 'A',\n",
    "        'GCA': 'A', 'GCG': 'A', 'UAU': 'Y', 'UAC': 'Y',\n",
    "        'UAA': '*', 'UAG': '*', 'UGA': '*', 'CAU': 'H',\n",
    "        'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q', 'AAU': 'N',\n",
    "        'AAC': 'N', 'AAA': 'K', 'AAG': 'K', 'GAU': 'D',\n",
    "        'GAC': 'D', 'GAA': 'E', 'GAG': 'E', 'UGU': 'C',\n",
    "        'UGC': 'C', 'UGG': 'W', 'CGU': 'R', 'CGC': 'R',\n",
    "        'CGA': 'R', 'CGG': 'R', 'AGA': 'R', 'AGG': 'R',\n",
    "        'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "    }\n",
    "\n",
    "    # Remove spaces from sequence\n",
    "    sequence = sequence.replace(\" \", \"\")\n",
    "\n",
    "    sequence = sequence.upper()\n",
    "\n",
    "    # Check for invalid characters\n",
    "    for char in sequence:\n",
    "        if char not in 'ATGCU':\n",
    "            return \"Error: Invalid character found. Only 'A', 'T', 'G', 'C', 'U' are allowed.\"\n",
    "\n",
    "    # Detect if it's DNA or RNA\n",
    "    if 'T' in sequence:\n",
    "        sequence = sequence.replace('T', 'U')  # Convert DNA to RNA\n",
    "\n",
    "    protein = \"\"\n",
    "\n",
    "    # Translate codons into amino acids\n",
    "    for i in range(0, len(sequence) , 3):\n",
    "        codon = sequence[i:i+3]\n",
    "        amino_acid = codon_table.get(codon, '?')  # '?' for unknown codons\n",
    "        if amino_acid == '*':  # Stop translation at stop codon\n",
    "            break\n",
    "        protein += amino_acid\n",
    "\n",
    "    return protein\n",
    "\n",
    "\n",
    "\n",
    "# Get user input\n",
    "user_seq = input(\"Enter a DNA or RNA sequence: \").strip().upper()\n",
    "\n",
    "# Translate and display the result\n",
    "translated_seq = translate_sequence(user_seq)\n",
    "print(\"\\nTranslated Protein Sequence:\", translated_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "59617f03-7c9a-4db6-bb59-828cb66ac9fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter a DNA or RNA sequence:  ACTCCGGACTTTGGCCTTAGCAGTAGTTAGTGTGGGAGGCTGGGAAGACTGGGAGCAGTCTCTTAAACAA AAGCAAAAGAATAAGCTTCGGGCGCTGTAGTACCTGCCAGCTTTCGCCACAGGAGCCTAGGCTGCCACCT GTAGGTCACTTGGGCTCCAGCTATGTGGCTGCCTCTTCTGCTGGGTGCCTTACTCTGGGCAGTGCTGTGG TTGCTCAGGGACCGGCAGAGCCTGCCCGCCAGCAATGCCTTTGTCTTCATCACCGGCTGTGACTCAGGCT TTGGGCGCCTTCTGGCACTGCAGCTGGACCAGAGAGGCTTCCGAGTCCTGGCCAGCTGCCTGACCCCCTC CGGGGCCGAGGACCTGCAGCGGGTGGCCTCCTCCCGCCTCCACACCACCCTGTTGGATATCACTGATCCC CAGAGCGTCCAGCAGGCAGCCAAGTGGGTGGAGATGCACGTTAAGGAAGCAGGGCTTTTTGGTCTGGTGA ATAATGCTGGTGTGGCTGGTATCATCGGACCCACACCATGGCTGACCCGGGACGATTTCCAGCGGGTGCT GAATGTGAACACAATGGGTCCCATCGGGGTCACCCTTGCCCTGCTGCCTCTGCTGCAGCAAGCCCGGGGC CGGGTGATCAACATCACCAGCGTCCTGGGTCGCCTGGCAGCCAATGGTGGGGGCTACTGTGTCTCCAAAT TTGGCCTGGAGGCCTTCTCTGACAGCCTGAGGCGGGATGTAGCTCATTTTGGGATACGAGTCTCCATCGT GGAGCCTGGCTTCTTCCGAACCCCTGTGACCAACCTGGAGAGTCTGGAGAAAACCCTGCAGGCCTGCTGG GCACGGCTGCCTCCTGCCACACAGGCCCACTATGGGGGGGCCTTCCTCACCAAGTACCTGAAAATGCAAC AGCGCATCATGAACCTGATCTGTGACCCGGACCTAACCAAGGTGAGCCGATGCCTGGAGCATGCCCTGAC TGCTCGACACCCCCGAACCCGCTACAGCCCAGGTTGGGATGCCAAGCTGCTCTGGCTGCCTGCCTCCTAC CTGCCAGCCAGCCTGGTGGATGCTGTGCTCACCTGGGTCCTTCCCAAGCCTGCCCAAGCAGTCTACTGAA TCCAGCCTTCCAGCAAGAGATTGTTTTTCAAGGACAAGGACTTTGATTTATTTCTGCCCCCACCCTGGTA CTGCCTGGTGCCTGCCACAAAATAAGCACTAACAAAAGTGTATTGTTTAAAAAATAAAAAGAAGGTGGGC AGAAATGTGCCCAGTGGAA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Single-frame Translation:\n",
      "Translated Protein Sequence: TPDFGLSSS\n",
      "\n",
      "Translated Protein Sequences from all Reading Frames:\n",
      "Forward Frame 1: MNLICDPDLTKVSRCLEHALTARHPRTRYSPGWDAKLLWLPASYLPASLVDAVLTWVLPKPAQAVY\n",
      "Forward Frame 2: MPLSSSPAVTQALGAFWHCSWTREASESWPAA\n",
      "Forward Frame 3: MADPGRFPAGAECEHNGSHRGHPCPAASAAASPGPGDQHHQRPGSPGSQWWGLLCLQIWPGGLL\n",
      "Reverse Frame 1: MVWVR\n",
      "Reverse Frame 2: MLQASAHLG\n",
      "Reverse Frame 3: MRCCIFRYLVRKAPP\n"
     ]
    }
   ],
   "source": [
    "def reverse_complement(seq):\n",
    "    \"\"\"Generate the reverse complement of an RNA sequence.\"\"\"\n",
    "    complement = {'A': 'U', 'U': 'A', 'G': 'C', 'C': 'G'}\n",
    "    return \"\".join(complement[base] for base in reversed(seq))\n",
    "\n",
    "def validate_sequence(sequence):\n",
    "    \"\"\"Check if the sequence contains only valid nucleotides (A, T, G, C, U).\"\"\"\n",
    "    valid_bases = {'A', 'T', 'G', 'C', 'U'}\n",
    "    for base in sequence:\n",
    "        if base not in valid_bases:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "def translate_sequence(sequence):\n",
    "    \"\"\"Translate a DNA or RNA sequence into a protein sequence.\"\"\"\n",
    "    codon_table = {\n",
    "        'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L',\n",
    "        'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',\n",
    "        'AUU': 'I', 'AUC': 'I', 'AUA': 'I', 'AUG': 'M',\n",
    "        'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',\n",
    "        'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',\n",
    "        'AGU': 'S', 'AGC': 'S', 'CCU': 'P', 'CCC': 'P',\n",
    "        'CCA': 'P', 'CCG': 'P', 'ACU': 'T', 'ACC': 'T',\n",
    "        'ACA': 'T', 'ACG': 'T', 'GCU': 'A', 'GCC': 'A',\n",
    "        'GCA': 'A', 'GCG': 'A', 'UAU': 'Y', 'UAC': 'Y',\n",
    "        'UAA': '*', 'UAG': '*', 'UGA': '*', 'CAU': 'H',\n",
    "        'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q', 'AAU': 'N',\n",
    "        'AAC': 'N', 'AAA': 'K', 'AAG': 'K', 'GAU': 'D',\n",
    "        'GAC': 'D', 'GAA': 'E', 'GAG': 'E', 'UGU': 'C',\n",
    "        'UGC': 'C', 'UGG': 'W', 'CGU': 'R', 'CGC': 'R',\n",
    "        'CGA': 'R', 'CGG': 'R', 'AGA': 'R', 'AGG': 'R',\n",
    "        'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "    }\n",
    "\n",
    "    # Remove spaces and convert to uppercase\n",
    "    sequence = sequence.replace(\" \", \"\").upper()\n",
    "\n",
    "    # Validate sequence\n",
    "    if not validate_sequence(sequence):\n",
    "        return \"Error: Invalid character found. Only 'A', 'T', 'G', 'C', 'U' are allowed.\"\n",
    "\n",
    "    # Convert DNA to RNA if necessary\n",
    "    if 'T' in sequence:\n",
    "        sequence = sequence.replace('T', 'U')\n",
    "\n",
    "    protein = \"\"\n",
    "\n",
    "    # Translate codons into amino acids\n",
    "    for i in range(0, len(sequence) - 2, 3):\n",
    "        codon = sequence[i:i+3]\n",
    "        amino_acid = codon_table.get(codon, '?')  # '?' for unknown codons\n",
    "        if amino_acid == '*':  # Stop translation at stop codon\n",
    "            break\n",
    "        protein += amino_acid\n",
    "\n",
    "    return protein\n",
    "\n",
    "def translate_frames(sequence):\n",
    "    \"\"\"Find all six reading frames and their translated protein sequences.\"\"\"\n",
    "    codon_table = {\n",
    "        'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L',\n",
    "        'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',\n",
    "        'AUU': 'I', 'AUC': 'I', 'AUA': 'I', 'AUG': 'M',\n",
    "        'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',\n",
    "        'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',\n",
    "        'AGU': 'S', 'AGC': 'S', 'CCU': 'P', 'CCC': 'P',\n",
    "        'CCA': 'P', 'CCG': 'P', 'ACU': 'T', 'ACC': 'T',\n",
    "        'ACA': 'T', 'ACG': 'T', 'GCU': 'A', 'GCC': 'A',\n",
    "        'GCA': 'A', 'GCG': 'A', 'UAU': 'Y', 'UAC': 'Y',\n",
    "        'UAA': '*', 'UAG': '*', 'UGA': '*', 'CAU': 'H',\n",
    "        'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q', 'AAU': 'N',\n",
    "        'AAC': 'N', 'AAA': 'K', 'AAG': 'K', 'GAU': 'D',\n",
    "        'GAC': 'D', 'GAA': 'E', 'GAG': 'E', 'UGU': 'C',\n",
    "        'UGC': 'C', 'UGG': 'W', 'CGU': 'R', 'CGC': 'R',\n",
    "        'CGA': 'R', 'CGG': 'R', 'AGA': 'R', 'AGG': 'R',\n",
    "        'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "    }\n",
    "\n",
    "    # Remove spaces and convert to uppercase\n",
    "    sequence = sequence.replace(\" \", \"\").upper()\n",
    "\n",
    "    # Validate sequence\n",
    "    if not validate_sequence(sequence):\n",
    "        return \"Error: Invalid character found. Only 'A', 'T', 'G', 'C', 'U' are allowed.\"\n",
    "\n",
    "    # Convert DNA to RNA if necessary\n",
    "    if 'T' in sequence:\n",
    "        sequence = sequence.replace('T', 'U')\n",
    "\n",
    "    # Generate 3 forward reading frames\n",
    "    forward_frames = [sequence[i:] for i in range(3)]\n",
    "    \n",
    "    # Generate 3 reverse complement reading frames\n",
    "    reverse_seq = reverse_complement(sequence)\n",
    "    reverse_frames = [reverse_seq[i:] for i in range(3)]\n",
    "\n",
    "    all_frames = forward_frames + reverse_frames\n",
    "    frame_labels = [\"Forward Frame 1\", \"Forward Frame 2\", \"Forward Frame 3\",\n",
    "                    \"Reverse Frame 1\", \"Reverse Frame 2\", \"Reverse Frame 3\"]\n",
    "\n",
    "    results = {}\n",
    "\n",
    "    # Translate each frame\n",
    "    for label, frame in zip(frame_labels, all_frames):\n",
    "        protein = \"\"\n",
    "        start_found = False  # Track if start codon (AUG) has been found\n",
    "\n",
    "        for i in range(0, len(frame) - 2, 3):\n",
    "            codon = frame[i:i+3]\n",
    "\n",
    "            if codon == 'AUG':  # Start translation at AUG\n",
    "                start_found = True\n",
    "                protein = \"M\"\n",
    "\n",
    "            elif start_found:\n",
    "                if codon in {'UAA', 'UAG', 'UGA'}:  # Stop translation at stop codon\n",
    "                    break\n",
    "                protein += codon_table.get(codon, '?')\n",
    "\n",
    "        results[label] = protein if protein else \"No valid protein found\"\n",
    "\n",
    "    return results\n",
    "\n",
    "# Get user input\n",
    "user_seq = input(\"Enter a DNA or RNA sequence: \").strip()\n",
    "\n",
    "print(\"\\nSingle-frame Translation:\")\n",
    "print(\"Translated Protein Sequence:\", translate_sequence(user_seq))\n",
    "\n",
    "print(\"\\nTranslated Protein Sequences from all Reading Frames:\")\n",
    "translated_frames = translate_frames(user_seq)\n",
    "for frame, protein_seq in translated_frames.items():\n",
    "    print(f\"{frame}: {protein_seq}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
